### PR TITLE
Don't slash all outgoing members.

### PR DESCRIPTION
**Pull request by kianenigma**
A small follow up to https://github.com/paritytech/substrate/pull/7384, this is not a bug per se but the general slashing attitude of the module was still a bit too aggressive. A members who is now downgraded to a runner-up should not yet be slashed. Instead, only those that are entirely ejected (no longer member + no longer runner-up) should be slashed. 

This could also cause members who bounced back and forth between being a member and runner to lose their `CandidacyBond`.

### CODE REVIEW

**Comment by thiolliere** [comment]
but that means there is currently a bug ?

because if a member get downgraded to runners-up and then back to member and then downgraded again it will be slashed 2 times, while having deposited only once.

If I'm correct then this issue should prioritised as High

**Comment by kianenigma** [comment]
Yes this is the behaviour, and indeed let's try and also merge this for the upcoming upgrade.

**Review by thiolliere — DISMISSED** [review]

**Comment by thiolliere** [comment]
also it might deserve a runtime note as well, I mean if someone get slashed_reserved too much and then fail to unreserve to funds unrelated to phragmen they could legitimetly ask for refund as well.

**Comment by kianenigma** [comment]
The reasoning for my wording is just that when I coded this initially, it was my intention to be a very strict module where losing any role would cause you to lose your bond. But now I see this is not really sensible.

**Review by shawntabrizi — APPROVED** [review]
Looks good to me.

**Comment by shawntabrizi** [comment]
I want to make sure I fully understand this. Users moving between `runner-up` and not in the list would not lose anything ,right? Because they would not be part of the "outgoing" set.

**Comment by thiolliere** [comment]
> I want to make sure I fully understand this. Users moving between `runner-up` and not in the list would not lose anything, right? Because they would not be part of the "outgoing" set.

they will be slashed reserved CandidacyDeposit, as to_burn_bond is extended with outgoing runners up

**Comment by thiolliere** [comment]
ah sorry I misclicked I want to rerequest a review for myself :-/

**Comment by thiolliere** [comment]
@kianenigma  to_burn_bond is extended without outgoing runners-up, but a candidate moving from runners-up to member is an outgoing runners-up, will he get incorrectly slashed ?

**Comment by shawntabrizi** [comment]
> they will be slashed reserved CandidacyDeposit, as to_burn_bond is extended with outgoing runners up

Then this seems just as bad then? Someone on the border is going to get slashed.

I think the only people who should be slashed are current members who actually completely leave the set.

In fact, I probably would not build in any slashing mechanism as part of this whole process, but rather build an explicit external mechanism to trigger a slash on a user. It is not clear that the behavior of dropping out of any set is actually the result of some malicious acting

**Comment by shawntabrizi** [comment]
@kianenigma I think you should remove this line:

https://github.com/paritytech/substrate/pull/7394/files#diff-194656be9ce9133b248a5b7cd18842eb5d3faacde3f1221b32e654419870136aR991

**Comment by shawntabrizi** [comment]
And in-fact you do not track the amount being slashed is only from the reserved balance allocated for this pallet. You can easily slash into other pallet's reserved balances.

I would strongly recommend that you remove the slashing mechanism for now until it is properly tracked and handled.

**Comment by thiolliere** [comment]
> And in-fact you do not track the amount being slashed is only from the reserved balance allocated for this pallet. You can easily slash into other pallet's reserved balances.
> 
> I would strongly recommend that you remove the slashing mechanism for now until it is properly tracked and handled.

Yes its planned in https://github.com/paritytech/substrate/pull/7040/files

**Comment by kianenigma** [comment]
I agree, the slashing here is too aggressive. I agree that outgoing members and runners-up should not be slashed. 

Should we keep the slash for loser candidates? that one is more sensible. It discourages any small account to give a shot at being a council member. The point is to only submit a candidacy when you think you are likely to win. 

But I guess this can be removed as well.

**Comment by shawntabrizi** [comment]
@kianenigma I think that one is fine. especially since the weight to submit candidacy grows with the number of candidates. You can have a reasonable attack to fill the beginning with a bunch of candidates, and force real candidates to pay larger fees to submit their candidacy.

**Comment by shawntabrizi** [comment]
@kianenigma sorry for so many follow ups here. I think i did not understand the pallet fully until I read more of the code.

I think actually the slash from runner up to nothing is fine because the user has the opportunity to renounce their candidacy at any time. They should do it or leave the system.

**Comment by thiolliere** [comment]
what about https://github.com/paritytech/substrate/pull/7394#issuecomment-716534749

from runners-up to member ppl get slashed_reserved too no ? should we fix here or in another follow up PR ?

**Comment by shawntabrizi** [comment]
> what about [#7394 (comment)](https://github.com/paritytech/substrate/pull/7394#issuecomment-716534749)
> 
> from runners-up to member ppl get slashed_reserved too no ? should we fix here or in another follow up PR ?

What i realized from reading the code more is that once they leave the `runner-ups` they are completely ejected from candidate pool. So there is no risk that someone might join `runner-ups` again without explicitly re-applying for candidacy. So there is no going back and forth in this scenario.

**Review by shawntabrizi — APPROVED** [review]

**Comment by thiolliere** [comment]
> > what about [#7394 (comment)](https://github.com/paritytech/substrate/pull/7394#issuecomment-716534749)
> > from runners-up to member ppl get slashed_reserved too no ? should we fix here or in another follow up PR ?
> 
> What i realized from reading the code more is that once they leave the `runner-ups` they are completely ejected from candidate pool. So there is no risk that someone might join `runner-ups` again without explicitly re-applying for candidacy. So there is no going back and forth in this scenario.

runners-up are candidate:
```rust
		let mut candidates = Self::candidates();
		// candidates who explicitly called `submit_candidacy`. Only these folks are at risk of
		// losing their bond.
		let exposed_candidates = candidates.clone();
		// current members are always a candidate for the next round as well.
		// this is guaranteed to not create any duplicates.
		candidates.append(&mut Self::members_ids());
		// previous runners_up are also always candidates for the next round.
		candidates.append(&mut Self::runners_up_ids());
```

**Comment by kianenigma** [comment]
@thiolliere is right this is also a case where we slash by mistake. You can be a runner and then be upgraded to a winner. This will cause you to be in the outgoing set of the runners-up and mistakingly slashed, while we shouldn't.

**Comment by kianenigma** [comment]
See the new test case that I added, it covers both cases and now I think we are finally good.

**Review by thiolliere — APPROVED** [review]

**Comment by kianenigma** [comment]
Much thanks to both of you for the rapid reviews ❤️

File: frame/elections-phragmen/src/lib.rs [hunk 1]
```diff
@@ -973,18 +973,29 @@ impl<T: Trait> Module<T> {
...
 			T::ChangeMembers::set_prime(prime);
 
-			// outgoing members lose their bond.
-			let mut to_burn_bond = outgoing.to_vec();
+			// outgoing members who are no longer a runner-up lose their bond.
+			let mut to_burn_bond = outgoing
+				.iter()
+				.filter(|o| new_runners_up_ids_sorted.binary_search(o).is_err())
+				.cloned()
+				.collect::<Vec<_>>();
```
Comment (apopiak) [inline frame/elections-phragmen/src/lib.rs:981]: looks like this will need re-benchmarking, right?

Comment (kianenigma) [reply frame/elections-phragmen/src/lib.rs:981]: Same as the other one, I think the change is insubstantial, but let's see

**Comment by kianenigma** [comment]
Insubstantial weight changes.

**Comment by kianenigma** [comment]
bot merge

**Comment by shawntabrizi** [comment]
bot merge

**Comment by kirushik** [comment]
bot merge

**Pull request merged**